File: java/src/Vigenere.java

import java.util.ArrayList;
import java.util.regex.Pattern;
import java.util.regex.Matcher;

/**
 * Created by postmordum on 6/29/15.
 */
public class Vigenere {
    String encrypt(String key,String text){
        int kp = 0;
        String encrypt = "";
        for(int i=0;i<text.length();i++){
            if(kp >= key.length()){
                kp = 0;
            }
            char t = text.charAt(i);
            char c = key.charAt(kp++);
            char e = this.shiftForward(c,t);
            //System.out.println(t+":"+c+":"+e);
            encrypt += e;
        }
        return encrypt;
    }
    String decrypt(String key,String text){
        int kp = 0;
        String decrypt = "";
        for(int i=0;i<text.length();i++){
            if(kp >= key.length()){
                kp = 0;
            }
            char t = text.charAt(i);
            char c = key.charAt(kp++);
            char d = this.shiftBackword(c,t);
            decrypt += d;
        }
        return decrypt;
    }
    private char shiftForward(char shift,char ltr){
        int shifted = ltr;
        if(ltr>31&&ltr<127){
            shifted = (ltr + shift);
            while(shifted > 126){
                int gap = shifted - 126;
                shifted = gap + 31;
            }
        }
        return (char)shifted;
    }
    private char shiftBackword(char shift,char ltr){
        int shifted = ltr;
        if(ltr>31&&ltr<127){
            shifted = (ltr - shift);
            while(shifted < 32){
                int gap = 32 - shifted;
                shifted = 127 - gap;
            }
        }
        return (char)shifted;
    }
}

File: python/vigenere.py



def shiftForward(shift,letter):
    shift_num = ord(shift)
    letter_num = ord(letter)
    shifted_num = letter_num
    if letter_num>31 and letter_num<127:
        shifted_num = (letter_num + shift_num)
        while shifted_num > 126:
            gap = shifted_num - 126
            shifted_num = gap + 31
    return chr(shifted_num)
    
def shiftBackward(shift,letter):
    shift_num = ord(shift)  
    letter_num = ord(letter)
    shifted_num = letter_num
    if letter_num>31 and letter_num<127:
        shifted_num = (letter_num - shift_num)
        while shifted_num < 32:
            gap = 32 - shifted_num
            shifted_num = 127 - gap
    return chr(shifted_num)
    
def vigenere(key,text,decrypt=None):
    if decrypt == None:
        decrypt = False
    output = "";
    key_len = len(key)
    for i,t in enumerate(text):
        kp = i%key_len;
        c = key[kp];
        if decrypt :
            o = shiftBackward(c,t)
        else:
            o = shiftForward(c,t)
        output += o;
    return output;

def test0():
    password = "abcd"
    text = "abcdefghijklmnopqrstuvwxyz"
    e = vigenere(password,text,False)
    d = vigenere(password,e,True)
    print("password:"+password)
    print("text:"+text)
    print("encrypted:"+e)
    success = (d==text)
    print("--PASS--" if success else "--FAIL--"+d)
    return success
def test1():
    password = "1243"
    text = "1234567890\nABCDEFGHIJKLMNOPQRSTUVWXYZ\n"
    e = vigenere(password,text,False)
    d = vigenere(password,e,True)
    print("password:"+password)
    print("text:"+text)
    print("encrypted:"+e)
    success = (d==text)
    print("--PASS--" if success else "--FAIL--"+d)
    return success
    
test0()
test1()
    
    
    
    
